xlators/features/index/src/index.c

```diff
@@ -425,23 +413,22 @@ static void
 index_update_link_count_cache(index_priv_t *priv, index_xattrop_type_t type,
                               int link_count_delta)
 {
+    int64_t ref = 0;
     switch (type) {
         case XATTROP:
-            LOCK(&priv->lock);
-            {
-                if (priv->pending_count >= 0) {
-                    if (link_count_delta == -1) {
-                        priv->pending_count--;
-                    }
-                    /*If this is the first xattrop, then pending_count needs to
-                     * be updated for the next lstat/lookup with link-count
-                     * xdata*/
-                    if (priv->pending_count == 0) {
-                        priv->pending_count--; /*Invalidate cache*/
-                    }
-                }
+            ref = GF_ATOMIC_GET(priv->pending_count);
+            if (ref >= 0) {
+                if (link_count_delta == -1)
+                    GF_ATOMIC_DEC(priv->pending_count);
```

Comment:
I'm unsure, but is that racy? Between line 419 and line 422, the value of pending_count can be changed (and actually be lower than 0) ?

Reply:
Yes, this code is unsafe.

Reply:
done